### ISSUE TITLE
no-undef: 2 with no-shadow: 0 is misreporting an issue

### ISSUE THREAD

**Issue by ronkorving** [eslint/eslint#3390]
Given the configuration:

- no-undef: 2
- no-shadow: 0

Should the following not be valid code?

```js
var Window = require('Window'); // Window is *not* window.Window
```

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by IanVS** [comment]
What is the actual ESLint output you get, and what version are you using?

**Comment by ronkorving** [comment]
Sorry, I should've pasted that in here of course:

```
   12:5   error    "Window" is read only        no-undef
```

**Comment by ronkorving** [comment]
FYI, I'm not running the require call and assignment in a function. It's at the top of my module.

**Comment by IanVS** [comment]
I'm guessing you have the `browser` environment set in your config file, right?  That is adding `Window` as a global read-only variable, and you're trying to overwrite it with your require.  If you really want to overwrite it, you should be able to add `/* global Window:true */` to the top of your file.

**Comment by ronkorving** [comment]
I have the browser environment, and because I'm using CommonJS, I'm not overwriting window.Window and I don't intend to either. I'm shadowing it within my module because the name happens to be the same as (window.)Window.

**Comment by mysticatea** [comment]
![image](https://cloud.githubusercontent.com/assets/1937871/9250955/8b3867be-420b-11e5-98dd-4d2c336a3894.png)

`no-undef` reports `Window` only in the global scope. And `var Window = 0;` rewrites `window. Window` in the global scope.

I guess you use a precompiler tool, so that is in the global scope, but that is modified into a module scope by tool. 
So is this an issue about combination of eslint and pre-compiler tools?

**Comment by ronkorving** [comment]
I don't use a precompiler. I lint my code as-is. But yes, "only in the global scope". How does it know if my module is a module (which it is) or actual global scope at SCA time? It doesn't, right?

**Comment by nzakas** [comment]
Currently, the only way to fix this is to use the `node` environment to add in the extra scope.

**Comment by ronkorving** [comment]
I guess. But the reason why I posted this issue is not to find a workaround. It's to find a proper long term solution that benefits the ESLint project and all of its users.

**Comment by nzakas** [comment]
This is related to #3377 and the solution is the same: add a commonjs environment that has the extra scope.

**Comment by ronkorving** [comment]
Sure, if that does the trick :+1:
